[Swift] Test simple cases of @_implementationOnly import

jrose-apple opened the pull request:
This feature deserves more tests, but this gets the "don't crash" part in place, and helped shake out early bugs on the compiler side. Continuing from #1414; I think it's ready to go now!

jrose-apple commented:
@swift-ci Please test

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/ReadMe.md:
```diff
@@ -0,0 +1,11 @@
+These tests concern the experimental Swift feature `@_implementationOnly import`.\* Unlike a normal import, an `@_implementationOnly` import is guaranteed not to be part of a library's API or ABI, which means that client apps don't have to know about the import at all to use the library. However, the import *can* be used in the *implementation* of the library, so it's important for LLDB to handle ...
```
An even better location for this text is as a docstring in the test function (or class) in the .py file. IIRC, the test driver knows about them, too.

jrose-apple replied:
This is one level up in a folder of four separate tests. I don't want to duplicate it in several places, but I mentioned it in the docstrings for the tests.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/library_indirect/Makefile:
```diff
@@ -0,0 +1,27 @@
+
+EXE=a.out
+SWIFT_SOURCES=main.swift
+SWIFTFLAGS_EXTRAS= -I$(shell pwd)
```
Does `-I. ` also work?

adrian-prantl replied:
Otherwise `Makefile.rules` defines `$(BUILDDIR)` as `$(shell pwd)`.

jrose-apple replied:
Oops, I forgot this one. I'll get it in the next rebase.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/library_indirect/TestLibraryIndirect.py:
```diff
@@ -0,0 +1,70 @@
+
+    mydir = TestBase.compute_mydir(__file__)
+    
+    @swiftTest
```
@JDevlieghere do we still use this decorator for anything?

JDevlieghere replied:
Yep, it ensures we skip everything but macOS/Linux and no i386.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/library_indirect/TestLibraryIndirect.py:
```diff
@@ -0,0 +1,70 @@
+        self.addTearDownHook(cleanup)
+        lldbutil.run_to_source_breakpoint(self, "break here", lldb.SBFileSpec("main.swift"))
+
+        self.expect("fr var", substrs=[
```
It may be less fragile to use self.frame().FindVariable("container") and then walk the value object tree programmatically instead of expecting the string output to match up perfectly. `TestObjCIvarDiscovery.py`  seems to have an example of doing that. There is also `lldbutil.check_variable()`.

jrose-apple replied:
I think you commented that last time, too. :-) I don't want that output; I want to see how the system accounts for the missing fields in how it's displayed to the user.

adrian-prantl commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/library_resilient/Makefile:
```diff
@@ -0,0 +1,27 @@
+
+EXE=a.out
+SWIFT_SOURCES=main.swift
+SWIFTFLAGS_EXTRAS= -I$(shell pwd)
```
ditto

adrian-prantl commented:
You can pass a dictionary with environment variables to make in `self.build()`. (For example `TestCppValueCast.py`). Is it feasibly to reuse source files and test functions more?

dcci commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/library_resilient/TestLibraryResilient.py:
```diff
@@ -0,0 +1,67 @@
+# TestLibrray.py
```
Small typo here.

dcci commented:
hmm, looks like this revealed a bug

```
14:01:22 AssertionError: False is not True : 'fr var' returns expected result, got 'value = <could not resolve type>
14:01:22 (main.ContainsTwoInts) container = (other = 10)
14:01:22 (Int) simple = 1'
```

jrose-apple commented:
Huh, it worked locally. I guess I need to rebase and rebuild.

jrose-apple commented:
> You can pass a dictionary with environment variables to make in self.build(). (For example TestCppValueCast.py). Is it feasibly to reuse source files and test functions more?

I can combine the two MainExecutable ones for sure. The Library ones are testing different things, though, even if they share a lot.

jrose-apple commented:
Hm. The macOS tests aren't failing locally for me, even after I rebase. Also, any idea what's up with the Linux failures? It sounds like the breakpoint's not being hit at all, but I'm not sure why that would be.

adrian-prantl commented:
> The macOS tests aren't failing locally for me, even after I rebase. 

Wild guess but: Maybe there is nondeterminism in the order in which modules are imported (something like as rdar://problem/50667488)?

> Also, any idea what's up with the Linux failures?

Maybe it's because 

`return // break here`

doesn't result in code that would suffice as a breakpoint location on Linux. Try

```
func stop() {}
stop() // break here
```

which is safer. If that isn't it, perhaps the Makefile somehow causes the Linux version to drop debug info?

jrose-apple commented:
Okay, I think I've figured out the difference in results: it's based on whether the *main* executable is using a dSYM or not. I'm not sure *why* that would affect the results, though; any idea?

adrian-prantl commented:
Perhaps it's the .swiftinterface files being copied into the dSYM that aren't found in the DWARF case?

jrose-apple commented:
@swift-ci Please test

jrose-apple committed:
[Swift] Test simple cases of @_implementationOnly import

This feature deserves more tests, but this gets the "don't crash" part
in place, and helped shake out early bugs on the compiler side.

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
Still stymied as to why this isn't working on Linux.

jrose-apple commented:
@swift-ci Please test Linux

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Still needs some re-review about some shortcuts I took.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
This PR seems to be cursed that only one platform can pass at a time.

@swift-ci Please test macOS

jrose-apple commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/main_executable/TestMainExecutable.py:
```diff
@@ -0,0 +1,147 @@
+        # (trying to debug something built against a library without that
+        # library present, implementation-only or not, and that library doesn't
+        # even have library evolution support enabled. Just make sure we don't
+        # crash.
```
@adrian-prantl @dcci Is this acceptable? I don't like it but I've already sunk way too much time into these few tests.

adrian-prantl replied:
They produce different results, but they are all different kinds of failures?
We *are* checking for the existence of any error, so I guess that is okay.

jrose-apple replied:
They are failing in different ways and it's really rather bizarre to me. I could file a Radar or Jira if you want it explicitly tracked somehow, but as noted it's not the most interesting configuration and I don't expect anyone to spend time tracking it down in practice.

adrian-prantl replied:
As long as we are testing for the fact *that they fail* that is good enough at the moment.

jrose-apple commented on packages/Python/lldbsuite/test/lang/swift/implementation_only_imports/library_resilient/TestLibraryResilient.py:
```diff
@@ -0,0 +1,88 @@
+        return info
+
+    @swiftTest
+    @expectedFailureOS(no_match(["macosx"])) # Requires Remote Mirrors support
```
@adrian-prantl @dcci Got a better idea for this condition?

adrian-prantl replied:
That condition LGTM.

jrose-apple commented:
Fixed the typo Davide spotted.

@swift-ci Please test
